### DISCUSSION TITLE
REPL instructions

### DISCUSSION THREAD

**Discussion by agzam** [agzam/spacehammer#128] [Show and tell]
@Grazfather @eccentric-j Could please describe how to get it running? It's been quite a while since I touched any Lua related stuff.  I naively thought that doing `luarock install jeejah` would  magically work, but then I remembered how Jay has mentioned - currently, you have to manually edit the rockspec. Can you list the steps required, prior to: 

    ;; - To your ~/.spacehammer/config.fnl add:
    ;;   (local repl (require :repl))
    ;;   (repl.run (repl.start))

**Comment by Grazfather** [comment]
You shouldn't need to do that anymore. Just `luarocks install jeejah`. After that you have to start the server and connect to it. I usually use [`shevek`](https://git.sr.ht/~technomancy/shevek).

I have the functionality bound as so:

```fnl
(local repl-keys
       [{:mods [:alt :cmd]
         :key :r
         :action (fn []
                   (alert "creating repl server")
                   (global repl-server (repl.start {:host "127.0.0.1"})))}
        {:mods [:alt :cmd]
         :key :s
         :action (fn []
                   (alert "running repl server")
                   (repl.run repl-server))}
        {:mods [:alt :cmd]
         :key :t
         :action (fn []
                   (alert "stopping repl server")
                   (repl.stop repl-server))}])
```

then
```bash
❯ fennel shevek.fnl localhost:7888
Connected.
>=> (+ 1 2)
3
```

**Comment by agzam** [comment]
Oh cool. At first I was seeing some errors in the log, so I assumed it didn't work. Then I was able to connect to it using Monroe. It starts the REPL and I even tried typing `(alert "I'm not a cow, I'm a pickle")` and it worked. 
Now I need to figure out how to connect to it from any Fennel file, so I can eval stuff in an Emacs buffer (not the Monroe REPL buffer).
But this is so cool. I love this!

**Comment by jaidetree** [comment]
@agzam Nice!

To do that open any fennel file and use `M-x monroe-interaction-mode` and `M-x monroe`, connect to the REPL server and you should be good to eval with `C-c C-c`. Just be mindful if you save a file and spacehammer restarts it will lose connection.

@Grazfather Thinking about that connection issue. The auto-reloading was likely a stop-gap due to not having a REPL. Thinking of wrapping that feature in a state machine then maybe whenever a client connects to the REPL server it will disable auto-reloading until a disconnect or manual restart.

**Comment by agzam** [comment]
> To do that open any fennel file and use M-x monroe-interaction-mode and M-x monroe, connect to the REPL server

That's what I tried. But I can't seem to find a way to connect to the instance. `M-x monroe` does open the REPL session. But monroe-interaction-mode doesn't do anything, except making some keys available

Here's what I did:

- in .spacehammer/config.fnl, I added these:

    (local repl (require :repl))
    (repl.run (repl.start {:port "9898"}))

- Then `M-x monroe`, it starts the REPL
- Attempting to connect to the REPL session from config.fnl, using `(monroe-connect "localhost:9898")`

  (error "Buffer config.fnl has no process")
  
- Attemping to enable monroe-interaction-mode and calling eval commands:

  (wrong-type-argument processp nil)

- Trying to check what the current session is `(monroe-current-session)`: 

  (wrong-type-argument processp nil)

**Reply by xiaobing-huang** [reply]
@agzam I also encountered the same problem, but now I managed to make it work.
check your .nrepl-port file, and see whether it is 9898, also there is an invisible character (carriage return) after 9898.
you need remove that also. I am use `echo -n "9898" > ~/Downloads/.nrepl-port` to remove the invisible character.

**Reply by agzam** [reply]
@xiaobing-huang Sorry for the long delay. Can you please elaborate how do you exactly connect to the REPL?
I am still having problems with associating any given .fnl file with REPL session. I can create a REPL session using `monroe` command. That allows me to type directly into the REPL, but I'd like to find a way of evaling things without having to type them in a REPL, but directly from .fnl files.

**Comment by tesujimath** [comment]
I spent a bit of time following these instructions.  [JeeJah](https://git.sr.ht/~technomancy/jeejah) has moved on, and is now native Fennel.  However, ultimately I gave up trying to get an nRepl server running inside Hammerspoon without it hogging the event loop.

In the end I simply installed a native Fennel REPL in Hammerspoon's console, following [A fennel REPL for Hammerspoon](https://www.linktohack.com/posts/a-fennel-repl-for-hammer-spoon/).  I had to change `fennel.repl` to `fennel.repl.repl` in `coroutine.create`, but otherwise it all Just Worked™.

So I added this near the bottom of my `config.fnl` just before it returns `config`.

```
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
;; Switch Hammerspoon console to use Fennel
;;
;; credit to https://www.linktohack.com/posts/a-fennel-repl-for-hammer-spoon/
;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;

(let [coro (coroutine.create fennel.repl.repl)]
    (coroutine.resume coro {:readChunk (fn []
                                         (let [input (coroutine.yield)]
                                           (.. input "\n")))
                            :onValues (fn [xs]
                                        (print (table.concat xs "\t")))
                            :onError (fn [_ msg]
                                       (print msg))})
    (set hs._consoleInputPreparser (fn [s]
                                     (coroutine.resume coro s)
                                     "")))
```

Posting here in case it helps someone else.

Thanks to all who have made Spacehammer great! ❤️
